### ISSUE TITLE
[enhancement] Move logger call into subroutine

### ISSUE THREAD

**Issue by feckert** [openwisp/openwrt-openwisp-monitoring#57]
The `logger` shell call is used over and over again.
Which makes the code confusing.
I would suggest that you move the call to a subroutine.

https://github.com/openwisp/openwrt-openwisp-monitoring/blob/master/openwrt-openwisp-monitoring/files/monitoring.init
https://github.com/openwisp/openwrt-openwisp-monitoring/blob/master/openwrt-openwisp-monitoring/files/monitoring.agent

**Comment by devkapilbansal** [comment]
@feckert Here subroutine refers to creating a specific function for logger calls?

**Comment by feckert** [comment]
@devkapilbansal In the described scripts in my description there is no function for the logger call.
Following the pattern don't repeat your self I would suggest that we add the following subroutine.
`log "<level>" "<message>"`
This makes the code clearer because it is one liner.

**Comment by nemesisdesign** [comment]
@devkapilbansal I have noticed that there are quite a few instances of the calls to `logger` in which the `-t` (tag) flag is not used. Is that on purpose?
I think all logger calls should have this.
I also second the proposal of @feckert, I think we can move the `-t openwisp_monitoring` to our function which we can then reuse. We should also write it so we don't have to pass "daemon.err" but only "err", moreover, we could default the log level to info so that in many cases the call to the logging facility will be very short.

I think this would make the code more readable since we have many calls to the logger.

**Comment by devkapilbansal** [comment]
Yes, I remove `-t` flag when logging in verbose mode. In verbose_mode, it is automatically tagging the logged message by the procd service name and if I pass a tag, then the message is logged twice

**Comment by nemesisdesign** [comment]
@devkapilbansal one more thing, regarding verbose mode, I think we could handle the verbose output using this function we're imagining here.

We have a lot of if/else in the code for verbose messages which are harming the readability of the agent code, what about solving this with our function?

Eg:

```
log "<message>" "<level>" "<verbose_message>" "<non_verbose_message_only>"
```

Practical examples below.

### Message shown regardless of mode

```
log "Something happened." "warn"
```

Result is shown both in verbose mode and non verbose:

```
Sep 18 18:18:20 Router daemon.warn openwisp-monitoring: Something happened.
```

### Message + verbose appended

```
log "Something happened." "warn" "Verbose output containing more detailed info here."
```

Result seen in non-verbose mode:

```
Sep 18 18:18:20 Router daemon.warn openwisp-monitoring: Something happened.
```

Result seen in verbose mode:

```
Sep 18 18:18:20 Router daemon.warn openwisp-monitoring: Something happened. Verbose output containing more detailed info here.
```

### Verbose message only

```
log "" "warn" "Message shown only in verbose mode."
```

Result seen only in verbose mode (in non-verbose mode the log line is not executed at all):

```
Sep 18 18:18:20 Router daemon.warn openwisp-monitoring: Message shown only in verbose mode.
```

### Non verbose message only

```
log "Something happened." "warn" "More details here..." "Run in verbose mode to find out more."
```

Non verbose result:

```
Sep 18 18:18:20 Router daemon.warn openwisp-monitoring: Something happened. Run in verbose mode to find out more.
```

Verbose result:

```
Sep 18 18:18:20 Router daemon.warn openwisp-monitoring: Something happened. More details here...
```

**Comment by nemesisdesign** [comment]
> Yes, I remove `-t` flag when logging in verbose mode. In verbose_mode, it is automatically tagging the logged message by the procd service name and if I pass a tag, then the message is logged twice

What do you think about adding a new "debug mode" to turn this on?
https://github.com/openwisp/openwrt-openwisp-monitoring/blob/bac6d545b0e3f21f2ca6d7e409220024963f111a/openwrt-openwisp-monitoring/files/monitoring.init#L89

In verbose mode we could avoid using it. In debug mode we would know (and document) that any output is sent to the logging facility, causing some output to be repeated, Or alternatively, we could even modify our log function to use echo instead of `logger`.

**Comment by devkapilbansal** [comment]
Can we do something like this :point_down: 

```
log -v -w "message"
```
```
-v --> Verbose mode
-w --> Warning
-e --> Error
-i --> Info
-n --> Non Verbose mode
```
When `-n` is passed, the message will be shown in both `verbose_mode` and `non-verbose mode`.
In the case of `-v`, the message will be shown in `verbose_mode` only.

**Comment by devkapilbansal** [comment]
Here are some examples on how it will work:

## Message shown regardless of mode
```
log -n -w "This is a warning"
```
The result will be shown in both verbose and non-verbose mode
```
Oct 14 03:46:15 Router daemon.warn openwisp-monitoring: This is a warning
```
### Verbose mode only
```
log -v -e "This is an error"
```
Result:
```
Oct 14 03:46:15 Router daemon.err openwisp-monitoring: This is an error
```

I don't take the other two cases into consideration as we are handling them in if-else to stop redundant messages in non-verbose mode as found [here](https://github.com/openwisp/openwrt-openwisp-monitoring/pull/34#pullrequestreview-733493335).
https://github.com/openwisp/openwrt-openwisp-monitoring/blob/5460f01df7053b8196edd1b3c092bb11ad557677/openwrt-openwisp-monitoring/files/monitoring.agent#L146-L157

**Comment by jedboywonder** [comment]
not sure if this is relevant, but I noticed openwisp2.log had some multi line logs  see bold text
this makes it harder to send to external log collector


2022-06-26 20:10:01.043 | syslog | localhost | /opt/openwisp2/log/openwisp2.log | [INFO 2022-06-27 03:10:00,633] module: views, process: 1616603, thread: 140152527804224
-- | -- | -- | -- | --
  | 2022-06-26 20:10:01.043 | syslog | localhost | /opt/openwisp2/log/openwisp2.log | Invalid data in "#/", validator says:
  | 2022-06-26 20:10:01.043 | syslog | localhost | /opt/openwisp2/log/openwisp2.log | None is not of type 'object'
  | 2022-06-26 20:10:01.043 | syslog | localhost | /opt/openwisp2/log/openwisp2.log | [WARNING 2022-06-27 03:10:00,634] module: log, process: 1616603, thread: 140152527804224
  | 2022-06-26 20:10:01.043 | syslog | localhost | /opt/openwisp2/log/openwisp2.log | Bad Request: /api/v1/monitoring/device/747e1fa36d7b4c98bb1f6b15a681ef99/
  **| 2022-06-26 20:10:01.043 | syslog | localhost | /opt/openwisp2/log/openwisp2.log |** [INFO 2022-06-27 03:10:00,760] module: views, process: 1616603, thread: 140152527804224
  | 2022-06-26 20:10:01.043 | syslog | localhost | /opt/openwisp2/log/openwisp2.log | **Invalid data in "#/", validator says:**
  | 2022-06-26 20:10:01.043 | syslog | localhost | /opt/openwisp2/log/openwisp2.log | **None is not of type 'object'**
  | 2022-06-26 20:10:01.043 | syslog | localhost | /opt/openwisp2/log/openwisp2.log | [**WARNING 2022-06-27 03:10**:00,761] module: log, process: 1616603, thread: 140152527804224
  | 2022-06-26 20:10:01.043 | syslog | localhost | /opt/openwisp2/log/openwisp2.log | Bad Request: /api/v1/monitoring/device/60b80f650bbe48dca079ef14dcb8615b/

**Comment by jedboywonder** [comment]
sry, I'm not good at github

![image](https://user-images.githubusercontent.com/105568928/175854173-29c907d8-ac88-4649-9ff6-b57ad9b37862.png)

**Comment by jedboywonder** [comment]
and...

It may be interesting to review the file permissions if the log stuff is being worked on.

root@localhost:/opt/openwisp2/log# ls -al *.log
-rwxrwx--- 1 root     www-data 24364516 Jun 27 03:25 celerybeat.log
-rwxrwx--- 1 root     www-data  3025201 Jun 21 04:53 celery-firmware-upgrader.log
-rw-r--r-- 1 root     root     19410196 Jun 27 03:25 celery.log
-rw-r--r-- 1 root     root     15458867 Jun 27 03:25 celery-monitoring.log
-rwxrwx--- 1 root     www-data  3134277 Jun 27 02:53 celery-network.log
-rwxrwx--- 1 root     www-data 15501336 Jun 27 02:41 daphne.log
-rw-r----- 1 www-data www-data  1471307 Jun 27 03:25 nginx.access.log
-rw-r----- 1 www-data www-data        0 Jun 22 00:00 nginx.error.log
-rw-r--r-- 1 www-data www-data   105600 Jun 27 03:25 openwisp2.log
-rw-r--r-- 1 root     root      1571985 Jun 27 03:25 uwsgi.log

Someone told me for Ubuntu, if you add a user to adm, then they should be able to see all the logs.
usermod -a -G adm humio-log-collector

For openwisp2, this is not the case.  There are some owned by root that are w:r so OK.  I just added this command to get the rest.
usermod -a -G www-data humio-log-collector

**Comment by nemesisdesign** [comment]
@jedboywonder these messages are out of topic here, I am going to delete them.

Django can be configure to send logs to a log collector. Use the general chat to ask questions.
